### PR TITLE
[Enhancement] Ngram bloom filter support like  predicate and case_insensitive mode

### PR DESCRIPTION
**Pull request by before-Sunrise**
## Why I'm doing:
ngram  bloom filter index support like predicate and "case_insensitive" property. "case_insensitive" property is for some  string function which is case insensitive like "ngram_search_case_insensitive"

for sql  in clickbench:
`"SELECT SearchPhrase, MIN(URL), MIN(Title), COUNT(*) AS c, COUNT(DISTINCT UserID) FROM hits WHERE Title LIKE '%Google%' AND URL NOT LIKE '%.google.%' AND SearchPhrase <> '' GROUP BY SearchPhrase ORDER BY c DESC LIMIT 10;"` using 1fe 3be(6core+32GB)

before this pr: 0.81s
after this pr with index on column "Title": 0.28s
`ALTER TABLE hits ADD INDEX idx_name1(Title) USING NGRAMBF ('gram_num' = "5", "bloom_filter_fpp" = "0.05");`

profile:

```
- BloomFilterFilter: 45.875ms
 - __MAX_OF_BloomFilterFilter: 57.738ms
 - __MIN_OF_BloomFilterFilter: 40.006ms
- BloomFilterFilterRows: 85.727M (85726654)
 - __MAX_OF_BloomFilterFilterRows: 3.648M (3648042)
 - __MIN_OF_BloomFilterFilterRows: 3.527M (3526633)
```

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/function_context.cpp [hunk 1]
```diff
@@ -22,6 +22,7 @@
 #include "column/type_traits.h"
 #include "exprs/agg/java_udaf_function.h"
 #include "runtime/runtime_state.h"
+#include "storage/rowset/bloom_filter.h"
```
Comment (kangkaisen) [inline be/src/exprs/function_context.cpp:25]: unused?

Comment (before-Sunrise) [reply be/src/exprs/function_context.cpp:25]: FunctionContext::~FunctionContext() = default; need to know NgramBloomFilterState's complete definition which is in bloom_filter.h

File: be/src/storage/schema_change_utils.cpp [hunk 2]
```diff
@@ -706,6 +706,10 @@ Status SchemaChangeUtils::parse_request_normal(const TabletSchemaCSPtr& base_sch
                        base_schema->has_index(ref_column.unique_id(), GIN)) {
                 *sc_directly = true;
                 return Status::OK();
+            } else if (new_schema->has_index(new_column.unique_id(), NGRAMBF) !=
+                       base_schema->has_index(ref_column.unique_id(), NGRAMBF)) {
+                *sc_directly = true;
```
Comment (kangkaisen) [inline be/src/storage/schema_change_utils.cpp:711]: do we need to consider the NgramBfIndexParams? such as gram num or bloom filter's false positive possibility

Comment (before-Sunrise) [reply be/src/storage/schema_change_utils.cpp:711]: right now ngram bloom filter only support add index and drop index, one column only have at most one ngram bloom filter index, so if want to modify gram num or fpp, user need to drop existed index and create new index

File: be/src/exprs/function_call_expr.cpp [hunk 3]
```diff
@@ -177,52 +177,53 @@ StatusOr<ColumnPtr> VectorizedFunctionCallExpr::evaluate_checked(starrocks::Expr
...
-            // not implemented right now
-            return true;
+    std::unique_ptr<NgramBloomFilterState>& ngram_state = fn_ctx->get_ngram_state();
+
+    // initialize ngram_state: determine whether this index useful or not, split needle into ngram_set if useful
+    if (ngram_state == nullptr) {
+        ngram_state = std::make_unique<NgramBloomFilterState>();
+        std::vector<Slice>& ngram_set = ngram_state->ngram_set;
+        bool index_useful;
+        if (_fn_desc->name == "LIKE") {
```
Comment (trueeyu) [inline be/src/exprs/function_call_expr.cpp:189]: Why change from like to LIKE?

File: be/src/exprs/function_call_expr.cpp [hunk 4]
```diff
@@ -177,52 +177,53 @@ StatusOr<ColumnPtr> VectorizedFunctionCallExpr::evaluate_checked(starrocks::Expr
...
+        for (auto& ngram : ngram_state->ngram_set) {
+            // if any ngram in needle doesn't hit bf, this page has nothing to do with target,so filter it
+            if (!bf->test_bytes(ngram.get_data(), ngram.get_size())) {
+                return false;
+            }
+        }
+        // if all ngram in needle hit bf, this page may have something to do with needle, so don't filter it
+        return true;
+    } else {
+        for (auto& ngram : ngram_state->ngram_set) {
```
Comment (trueeyu) [inline be/src/exprs/function_call_expr.cpp:218]: NOT LIKE and others?

File: be/src/exprs/function_call_expr.cpp [hunk 5]
```diff
@@ -232,8 +233,113 @@ bool VectorizedFunctionCallExpr::support_ngram_bloom_filter(ExprContext* context
         return false;
     }
 
-    return _fn_desc->name == "regex" || _fn_desc->name == "like" || _fn_desc->name == "ngram_search" ||
+    return _fn_desc->name == "LIKE" || _fn_desc->name == "ngram_search" ||
```
Comment (trueeyu) [inline be/src/exprs/function_call_expr.cpp:236]: why remove regex?

File: fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java [hunk 6]
```diff
@@ -45,6 +45,9 @@ public class FeConstants {
 
     // NGRAM Bloom filter's default gram number
     public static final int DEFAULT_GRAM_NUM = 2;
+
+    public static final String NGRAM_CASE_SENSITIVE = "true";
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java:49]: why not use boolean?

File: be/src/exprs/function_call_expr.cpp (hunk 3 above)
Comment (before-Sunrise) [reply be/src/exprs/function_call_expr.cpp:189]: [60010, 'LIKE', 'BOOLEAN', ['VARCHAR', 'VARCHAR'], 'LikePredicate::like', 'LikePredicate::like_prepare',  which cause function name is "LIKE"

File: be/src/exprs/function_call_expr.cpp (hunk 5 above)
Comment (before-Sunrise) [reply be/src/exprs/function_call_expr.cpp:236]: not implemented right now

File: be/src/util/slice.h [hunk 7]
```diff
@@ -182,6 +183,13 @@ class Slice {
         return ((size >= x.size) && memequal(data + (size - x.size), x.size, x.data, x.size));
     }
 
+    Slice tolower(std::string& buf) {
+        // copy this slice into buf
+        buf.assign(get_data(), get_size());
+        std::transform(buf.begin(), buf.end(), buf.begin(), [](unsigned char c) { return std::tolower(c); });
```
Comment (trueeyu) [inline be/src/util/slice.h:189]: Why not use std::transform(buf.begin(), buf.end(), buf.begin(), ::tolower);

File: be/src/storage/rowset/column_reader.cpp [hunk 8]
```diff
@@ -686,25 +687,32 @@ size_t ColumnReader::mem_usage() const {
...
+    NgramBloomFilterReaderOptions reader_options;
     std::shared_ptr<TabletIndex> ngram_bf_index;
 
     Status status = _segment->tablet_schema().get_indexes_for_column(_column_unique_id, NGRAMBF, ngram_bf_index);
     if (!status.ok() || ngram_bf_index.get() == nullptr) {
-        return gram_num;
+        return reader_options;
     }
 
     const std::map<std::string, std::string>& index_properties = ngram_bf_index->index_properties();
```
Comment (satanson) [inline be/src/storage/rowset/column_reader.cpp:700]: Is index_properties is a persistent string-to-string map? it seems use a bitvector to store some encoded flag is enough, just like in page.flags in Linux or cr0~3 register of intel cpu.

File: be/src/exprs/function_call_expr.cpp [hunk 9]
```diff
@@ -232,8 +233,113 @@ bool VectorizedFunctionCallExpr::support_ngram_bloom_filter(ExprContext* context
...
+    // checked in support_ngram_bloom_filter(size_t gram_num), so it 's safe to get const column's value
+    Slice needle;
+    const auto& needle_column = fn_ctx->get_constant_column(1);
+
+    if (index_case_sensitive) {
+        needle = ColumnHelper::get_const_value<TYPE_VARCHAR>(needle_column);
+    } else {
+        // for case_insensitive, we need to convert needle to lower case
+        std::string& buf = ngram_state->buffer;
+        needle = ColumnHelper::get_const_value<TYPE_VARCHAR>(needle_column).tolower(buf);
```
Comment (satanson) [inline be/src/exprs/function_call_expr.cpp:272]: is needle column always a constant ?

File: be/src/exprs/function_call_expr.cpp [hunk 10]
```diff
@@ -177,52 +177,53 @@ StatusOr<ColumnPtr> VectorizedFunctionCallExpr::evaluate_checked(starrocks::Expr
...
 
-    for (auto& ngram : ngram_set) {
-        if (bf->test_bytes(ngram.get_data(), ngram.get_size())) {
-            return true;
+    // if empty, which means needle is too short, so index_valid should be false
+    DCHECK(!ngram_state->ngram_set.empty());
+    if (_fn_desc->name == "LIKE") {
+        for (auto& ngram : ngram_state->ngram_set) {
+            // if any ngram in needle doesn't hit bf, this page has nothing to do with target,so filter it
+            if (!bf->test_bytes(ngram.get_data(), ngram.get_size())) {
```
Comment (satanson) [inline be/src/exprs/function_call_expr.cpp:211]: so the bf is constructed after the column has been to-lowercase-ed?

File: be/src/storage/rowset/bloom_filter_index_writer.cpp [hunk 11]
```diff
@@ -225,11 +220,16 @@ class NgramBloomFilterIndexWriterImpl<TYPE_VARCHAR> : public BloomFilterIndexWri
...
                 if (_values.find(unaligned_load<CppType>(&cur_ngram)) == _values.end()) {
-                    // why deep copy?
-                    _values.insert(get_value<TYPE_VARCHAR>(&cur_ngram, this->_typeinfo, &this->_pool));
+                    if (this->_bf_options.case_sensitive) {
+                        _values.insert(get_value<TYPE_VARCHAR>(&cur_ngram, this->_typeinfo, &this->_pool));
+                    } else {
+                        // todo::exist two copy of ngram, need to optimize
+                        std::string lower_ngram;
+                        Slice lower_ngram_slice = cur_ngram.tolower(lower_ngram);
+                        _values.insert(get_value<TYPE_VARCHAR>(&lower_ngram_slice, this->_typeinfo, &this->_pool));
```
Comment (satanson) [inline be/src/storage/rowset/bloom_filter_index_writer.cpp:229]: to_lower_case can be vectorized efficiently, suggest to reference to StringFunction::tolower

File: be/src/exprs/function_call_expr.cpp (hunk 10 above)
Comment (satanson) [inline be/src/exprs/function_call_expr.cpp:211]: so the bf is constructed after the column has been to-lowercase-ed?

File: be/src/exprs/function_call_expr.cpp [hunk 12]
```diff
@@ -232,8 +233,113 @@ bool VectorizedFunctionCallExpr::support_ngram_bloom_filter(ExprContext* context
...
+    std::vector<size_t> index;
+    size_t slice_gram_num = get_utf8_index(needle, &index);
+    // case like "ngram_search
+    if (slice_gram_num < index_gram_num) {
+        return false;
+    }
+    ngram_set.reserve(slice_gram_num - index_gram_num + 1);
+
+    size_t j;
+    for (j = 0; j + index_gram_num <= slice_gram_num; j++) {
```
Comment (trueeyu) [inline be/src/exprs/function_call_expr.cpp:284]: for (size_t j = 0; ?

**Review by trueeyu — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**